File: microbrain/backend/app/engine/platform_compilers.py
"""Platform-specific prompt compilers.

Each compiler takes the full KIMIFY-layered blueprint dict produced by
visual_blueprint.parse_blueprint() and formats it for the target platform,
mirroring the layer assembly order used by the frontend's promptBuilder.js
(Kernel -> Intent -> Medium -> Illumination -> Format) so Midjourney, DALL-E
and Nano Banana all apply the same prompt-engineering rules.
"""


def _layer_strings(blueprint: dict) -> dict:
    kernel_parts = [blueprint["full_description"]]
    if blueprint["subject_details"]:
        kernel_parts.append(", ".join(blueprint["subject_details"]))
    if blueprint["subject_modifiers"]:
        kernel_parts.append(", ".join(blueprint["subject_modifiers"]))

    intent_parts = [blueprint["style_intent"], blueprint["mood"], blueprint["visual_style"], blueprint["era_period"]]
    medium_parts = [blueprint["camera"], blueprint["lens"], blueprint["film"], blueprint["rendering_engine"]]
    illum_parts = [blueprint["lighting"], blueprint["time_of_day"], blueprint["atmosphere"], blueprint["weather"]]
    format_parts = [blueprint["composition"], blueprint["angle"], blueprint["dof"], blueprint["color"]]

    return {
        "kernel": ", ".join(p for p in kernel_parts if p),
        "intent": ", ".join(p for p in intent_parts if p),
        "medium": ", ".join(p for p in medium_parts if p),
        "illumination": ", ".join(p for p in illum_parts if p),
        "format": ", ".join(p for p in format_parts if p),
    }


def compile_midjourney(blueprint: dict, params: dict) -> str:
    layers = _layer_strings(blueprint)
    content = ", ".join(v for v in layers.values() if v)

    # Manual UI params only override the inferred domain values when the
    # user actually changed them from the platform-neutral defaults.
    ar = params.get("ar", "1:1")
    if ar == "1:1":
        ar = blueprint["aspect_ratio"]
    stylize = params.get("s", 100)
    if stylize == 100:
        stylize = blueprint["stylize"]
    chaos = params.get("c", 0)
    if chaos == 0:
        chaos = blueprint["chaos"]
    weird = params.get("w", 0)
    quality = params.get("q", 1)

    suffix = [f"--ar {ar}"]
    if stylize != 100:
        suffix.append(f"--s {stylize}")
    if chaos != 0:
        suffix.append(f"--c {chaos}")
    if weird != 0:
        suffix.append(f"--w {weird}")
    if blueprint["raw"]:
        suffix.append("--raw")
    if quality != 1:
        suffix.append(f"--q {quality}")
    suffix.append("--v 8.1")

    return f"{content} {' '.join(suffix)}"


def compile_dalle(blueprint: dict) -> str:
    layers = _layer_strings(blueprint)
    desc = layers["kernel"]
    article = "an" if desc[:1].lower() in "aeiou" else "a"

    sentences = [f"Create an image of {article} {desc}."]
    if layers["intent"]:
        sentences.append(f"{layers['intent'].capitalize()}.")
    if layers["medium"]:
        sentences.append(f"Captured with {layers['medium']}.")
    if layers["illumination"]:
        sentences.append(f"{layers['illumination'].capitalize()}.")
    if layers["format"]:
        sentences.append(f"{layers['format'].capitalize()}.")
    sentences.append("Highly detailed, professional quality.")

    return " ".join(sentences)


def compile_nano_banana(blueprint: dict) -> str:
    layers = _layer_strings(blueprint)
    desc = layers["kernel"]

    parts = [f"Transform into a cinematic scene: {desc}."]
    if layers["intent"]:
        parts.append(f"{layers['intent'].capitalize()}.")
    if layers["medium"]:
        parts.append(f"Shot using {layers['medium']}.")
    if layers["illumination"]:
        parts.append(f"{layers['illumination'].capitalize()} sets the atmosphere.")
    if layers["format"]:
        parts.append(f"Strong {layers['format']} composition, striking environmental detail.")

    return " ".join(parts)

File: microbrain/frontend/src/lib/platformCompilers.js
// ============================================================
// Unified platform compilers (frontend offline path)
// Reuses the same domain-detection + KIMIFY-layers engine that
// powers Pro Builder's auto-inference (kimifyInference.js), so the
// Generador's offline fallback applies the same prompt-engineering
// rules as the online backend path (mj81_compiler.py) instead of a
// trivial hardcoded template.
// ============================================================

import { inferFromSubject } from '@/lib/kimifyInference';

function layerStrings(subjectText, fields) {
  const kernelParts = [subjectText];
  if (fields.subjectDetails?.length) kernelParts.push(fields.subjectDetails.join(', '));
  if (fields.subjectModifiers?.length) kernelParts.push(fields.subjectModifiers.join(', '));

  const intentParts = [fields.styleIntent, fields.mood, fields.visualStyle, fields.eraPeriod].filter(Boolean);
  const mediumParts = [fields.cameraType, fields.lensType, fields.filmType, fields.renderingEngine].filter(Boolean);
  const illumParts = [fields.lightingType, fields.timeOfDay, fields.atmosphere, fields.weather].filter(Boolean);
  const formatParts = [fields.composition, fields.angle, fields.depthOfField, fields.colorGrading].filter(Boolean);

  return {
    kernel: kernelParts.filter(Boolean).join(', '),
    intent: intentParts.join(', '),
    medium: mediumParts.join(', '),
    illumination: illumParts.join(', '),
    format: formatParts.join(', '),
  };
}

function capitalize(s) {
  return s.charAt(0).toUpperCase() + s.slice(1);
}

function compileMidjourney(layers, fields, params) {
  const content = [layers.kernel, layers.intent, layers.medium, layers.illumination, layers.format]
    .filter(Boolean)
    .join(', ');

  // Manual UI params only override the inferred domain values when the
  // user actually changed them from the platform-neutral defaults.
  const ar = params.ar && params.ar !== '1:1' ? params.ar : fields.aspectRatio;
  const stylize = params.s && params.s !== 100 ? params.s : fields.stylize;
  const chaos = params.c && params.c !== 0 ? params.c : fields.chaos;
  const weird = params.w || 0;
  const quality = params.q || 1;

  const suffix = [`--ar ${ar}`];
  if (stylize !== 100) suffix.push(`--s ${stylize}`);
  if (chaos !== 0) suffix.push(`--c ${chaos}`);
  if (weird !== 0) suffix.push(`--w ${weird}`);
  if (fields.raw) suffix.push('--raw');
  if (quality !== 1) suffix.push(`--q ${quality}`);
  suffix.push('--v 8.1');

  return `${content} ${suffix.join(' ')}`;
}

function compileDalle(layers) {
  const desc = layers.kernel;
  const article = /^[aeiou]/i.test(desc) ? 'an' : 'a';

  const sentences = [`Create an image of ${article} ${desc}.`];
  if (layers.intent) sentences.push(`${capitalize(layers.intent)}.`);
  if (layers.medium) sentences.push(`Captured with ${layers.medium}.`);
  if (layers.illumination) sentences.push(`${capitalize(layers.illumination)}.`);
  if (layers.format) sentences.push(`${capitalize(layers.format)}.`);
  sentences.push('Highly detailed, professional quality.');

  return sentences.join(' ');
}

function compileNanoBanana(layers) {
  const desc = layers.kernel;

  const parts = [`Transform into a cinematic scene: ${desc}.`];
  if (layers.intent) parts.push(`${capitalize(layers.intent)}.`);
  if (layers.medium) parts.push(`Shot using ${layers.medium}.`);
  if (layers.illumination) parts.push(`${capitalize(layers.illumination)} sets the atmosphere.`);
  if (layers.format) parts.push(`Strong ${layers.format} composition, striking environmental detail.`);

  return parts.join(' ');
}

export function compilePlatformPrompt(inputText, platform, params = {}) {
  const base = (inputText || '').trim();
  if (!base) return null;

  const { fields, meta } = inferFromSubject(base);
  const layers = layerStrings(base, fields);

  let positive;
  if (platform === 'midjourney_v8_1') {
    positive = compileMidjourney(layers, fields, params);
  } else if (platform === 'dalle_3') {
    positive = compileDalle(layers);
  } else {
    positive = compileNanoBanana(layers);
  }

  const negative =
    platform === 'midjourney_v8_1' && params.output_mode === 'prompt_plus_negative' ? fields.negative : '';

  const canvas = negative ? `POSITIVE PROMPT\n${positive}\n\nNEGATIVE PROMPT\n${negative}` : positive;

  return {
    status: 'DONE_WITH_PROMPT',
    positive_prompt: positive,
    negative_prompt: negative,
    canvas,
    file: null,
    platform,
    domain: meta,
  };
}
